### PR TITLE
Add log target and l2met transform

### PR DESCRIPTION
**Pull request by stevenharman**
This builds atop [#20](https://github.com/babbel/puma-plugin-telemetry/pull/20). If/when it's merged, I'll rebase these changes atop `main` to make these changes more clear.

This PR adds a new `LogTarget`, and splits out `Formatters` and `Transforms` into their own objects, so they can be mixed and matched. The `LogTarget` defaults to using the Ruby stdlib `::Logger` pointed at `$stdout`, using an Logfmt `formatter`. We can then use the new L2Met transform to get the metrics in our logs, exactly as we need them!

```ruby
Puma::Plugin::Telemetry::Targets::LogTarget.new(transform: :l2met).call('queue.size': 42, load: 0.3)
#=> I, [2024-05-02T15:40:08.222977 #51368]  INFO -- : sample#queue.size=42 sample#load=0.3 name="Puma::Plugin::Telemetry"

# Or, JSON-ify them
Puma::Plugin::Telemetry::Targets::LogTarget.new(transform: :l2met, formatter: :json).call('queue.size': 42, load: 0.3)
#=> I, [2024-05-02T15:42:27.037245 #51368]  INFO -- : {"sample#queue.size":42,"sample#load":0.3,"name":"Puma::Plugin::Telemetry"}
```

These also work with the `IOTarget`:

```ruby
Puma::Plugin::Telemetry::Targets::IOTarget.new(transform: :l2met, formatter: :logfmt).call('queue.size': 42, load: 0.3)
#=> sample#queue.size=42 sample#load=0.3 name="Puma::Plugin::Telemetry"
```

The default `formatter:` and `transform:` options for `IOTarget` keep it backward compatible with the way that target worked before.

I've tried to make each step along the way an atomic commit. It might be helpful to quickly look commit-by-commit if you're curious. Or not - most of the churn is just reorganizing the code that was here.

Phew! Thanks so much for this tooling, and for your consideration.

### CODE REVIEW

File: lib/puma/plugin/telemetry/transforms/l2met_transform.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  class Plugin
+    module Telemetry
+      module Transforms
+        # L2Met (Logs to Metrics) transform that makes all keys a `sample#` in the L2Met format.
+        #
+        class L2metTransform
+          def self.call(telemetry)
+            telemetry.transform_keys { |k| "sample##{k}" }.tap do |data|
+              data['name'] = 'Puma::Plugin::Telemetry'
+              # data['source'] = ??
```
Comment (stevenharman) [inline lib/puma/plugin/telemetry/transforms/l2met_transform.rb:13]: L2Met allows a special "source" key, and I'm not sure of the best way to handle that. For example, if deployed on Heroku, you could image `data['source'] ||= ENV.fetch('DYNO', nil)` or something like that.

Comment (stevenharman) [reply lib/puma/plugin/telemetry/transforms/l2met_transform.rb]: I ended up adding in a simple mechanism that will fetch `L2MET_SOURCE` or `DYNO` from `ENV`, falling back to `$PROGRAM_NAME`. Seems a sensible place to start.

File: lib/puma/plugin/telemetry/transforms/l2met_transform.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+              data['source'] ||= source
+            end
+          end
+
+          private
+
+          attr_reader :host_env, :program_name, :socket
+
+          def source
+            host_env['L2MET_SOURCE'] ||
```
Comment (driv3r) [inline lib/puma/plugin/telemetry/transforms/l2met_transform.rb:34]: `source` looks like a perfect thing to cache under instance variable

**Comment by driv3r** [comment]
I'll try to properly review tomorrow morning 👍

**Comment by unsign3d** [comment]
This is a bit of a big PR, I will quickly create a ticket for ourselves to go through it and test it.
Thank you for contributing to this tool :)

**Comment by unsign3d** [comment]
I just realised that I removed @driv3r from the reviewer.
Feel free to review, that was an accident (and I kinda didn't even understand how I did it) XD

File: README.md [hunk 3]
```diff
@@ -41,22 +41,61 @@ Puma::Plugin::Telemetry.configure do |config|
...
+This target has configurable `formatter:` and `transform:` options.
+The `formatter:` options are
+
+* `:json` _(default)_ - Print the logs in JSON.
+* `:logfmt` - Print the logs in key/value pairs, as per `logfmt`.
+* `:noop` - A NOOP formatter which returns the telemetry `Hash` unaltered, passing it directly to the `io:` instance.
+
+The `transform:` options are
+
+* `:cloud_watch` _(default)_ - Transforms telemetry keys, replacing dots with dashes to support AWS CloudWatch Log Metrics filters.
```
Comment (stevenharman) [inline README.md:63]: This does keep things backward compatible, though I wonder if it's a bit surprising and changing the default, as part of a breaking-change 2.0 release, might be the way to go?

I could see the `:noop` transform being a sensible default. Then only if you're using CloudWatch Log Metrics, or L2Met, would you need to opt for a different transform. Thoughts?

File: README.md [hunk 4]
```diff
@@ -41,22 +41,61 @@ Puma::Plugin::Telemetry.configure do |config|
...
 ```
 
+#### Options
+
+This target has configurable `formatter:` and `transform:` options.
+The `formatter:` options are
+
+* `:json` _(default)_ - Print the logs in JSON.
+* `:logfmt` - Print the logs in key/value pairs, as per `logfmt`.
+* `:noop` - A NOOP formatter which returns the telemetry `Hash` unaltered, passing it directly to the `io:` instance.
```
[the remark this answers is not in the document]
Comment (stevenharman) [reply README.md:59]: I don't really love it either. `:passthrough` seems a reasonable alternative to me. Anyone have other alternatives or thoughts?

[the remark this answers is not in the document]
Comment (stevenharman) [reply README.md:59]: Any further thoughts? I'm happy to rename to `:passthrough`; it is a bit more explicit of _how_ it's not doing anything. 😄

**Comment by unsign3d** [comment]
First of all, thanks again for the contribution 👊 

I really like the modularity and concepts you've introduced here. However, I feel this PR is bringing in quite a few changes, and I'm not fully sold on some of them yet.

If you're open for it, I would propose you to split this PR into 3 separate PRs:

- A PR that does the refactoring and exposes the noop classes as base classes, I feel that this would bring major advantages to people if they want to create new formatters and transformers, in this case you could leverage our codebase without needing to fork necessarily
- A PR that adds the logfmt and logger as possible options, based on the base classes (I feel this would be a good example that people could build on top)
- A PR that adds the l2met transform based on the base class, and check if it makes more sense as a plugin or as an upstream option

I think this contribution would make it for an awesome base, and people could start using their transformers without forking the code, and without constant rebasing, therefore reducing overall maintenance.

Does this approach make sense to you? I'm happy to discuss further!

**Comment by stevenharman** [comment]
@unsign3d yeah, if you'd rather see it broken up (1 & 2) and then potentially pass on the `l2met` formatter (3), that's fine. I'll break it up and open new PRs.

FWIW, if this was accepted my plan was to then work on adding the right amount of `autoload` so these extra transforms and formatters were only loaded into memory if/when they were used. That way having some extra ones (like `l2met`) wouldn't really have any impact to users.

**Comment by unsign3d** [comment]
@stevenharman that sounds good :) 
What I'm not convinced is the maintenance of formatters after they become mainline, especially the l2met that we are not using and I'm not really sure how it works... 
I'm leaning towards the idea of having another repo / gem with that formatter as a plugin, so that it can be developed independently and with more knowledge of the l2met platform that my team currently has 🤔

**Comment by stevenharman** [comment]
Happy to take this to a different medium (GitHub Discussions?), but the gist on L2Met is it was primarily created/evolved as a way to [emit metrics via log streams](https://github.com/ryandotsmith/l2met/blob/master/README.md). The largest use case I know of is Librato (now SolarWinds), and it's pretty much the current standard for anything deployed to Heroku. That happens to be where all of our stuff is deployed, and we're currently using Librato. Hence, the need/desire for these logs to be emitted in [L2Met format](https://github.com/ryandotsmith/l2met/wiki/Usage).

Does it belong in the main repo? I don't have a particularly strong opinion either way. That said, it's extremely stable (and pretty much in maintenance mode), and I'd expect very little carrying cost for this repo. It's also such a simple `transform` that it's trivial to extract. So… 🤷

**Comment by unsign3d** [comment]
Let's start by pushing it to upstream, and then if it becomes too much maintenance we can create a new major and move it into a plugin :)

**Comment by stevenharman** [comment]
Closing in favor of #25 and #26.

**Pull request closed without merging**